pkg/operator/reconcile.go

```diff
@@ -89,36 +94,88 @@ const (
 )
 
 // MergeDeploymentForUpdate updates Deployment objects.
-// The existing revision annotation needs to take precedence
+// We merge annotations, keeping ours except the Deployment Revision annotation.
 func MergeDeploymentForUpdate(current, updated *uns.Unstructured) error {
-	if (updated.GetAPIVersion() == "apps/v1" || updated.GetAPIVersion() == "v1beta2") && updated.GetKind() == "Deployment" {
+	gvk := updated.GroupVersionKind()
+	if gvk.Group == "apps" && gvk.Kind == "Deployment" {
+
+		// Copy over the revision annotation from current up to updated
+		// otherwise, updated would win, and this annotation is "special" and
+		// needs to be preserved
+		curAnnotations := current.GetAnnotations()
 		updatedAnnotations := updated.GetAnnotations()
 		if updatedAnnotations == nil {
 			updatedAnnotations = map[string]string{}
 		}
-		curAnnotations := current.GetAnnotations()
 
 		if curAnnotations != nil {
-			updatedAnnotations[deploymentRevisionAnnotation] = curAnnotations[deploymentRevisionAnnotation]
-			updated.SetAnnotations(updatedAnnotations)
+			anno, ok := curAnnotations[deploymentRevisionAnnotation]
+			if ok {
+				updatedAnnotations[deploymentRevisionAnnotation] = anno
+			}
 		}
+
+		updated.SetAnnotations(updatedAnnotations)
 	}
 
 	return nil
 }
 
 // MergeServiceForUpdate ensures the clusterip is never written to
 func MergeServiceForUpdate(current, updated *uns.Unstructured) error {
-	if updated.GetAPIVersion() == "v1" && updated.GetKind() == "Service" {
-		clusterIP, found, err := unstructured.NestedString(current.Object, "spec", "clusterIP")
+	gvk := updated.GroupVersionKind()
+	if gvk.Group == "" && gvk.Kind == "Service" {
+		clusterIP, found, err := uns.NestedString(current.Object, "spec", "clusterIP")
 		if err != nil {
 			return err
 		}
 
 		if found {
-			return unstructured.SetNestedField(updated.Object, clusterIP, "spec", "clusterIP")
+			return uns.SetNestedField(updated.Object, clusterIP, "spec", "clusterIP")
 		}
 	}
 
 	return nil
 }
+
+// mergeAnnotations copies over any annotations from current to updated,
+// with updated winning if there's a conflict
+func mergeAnnotations(current, updated *uns.Unstructured) {
+	updatedAnnotations := updated.GetAnnotations()
+	curAnnotations := current.GetAnnotations()
+
+	if curAnnotations == nil {
+		curAnnotations = map[string]string{}
+	}
+
+	if updatedAnnotations == nil {
+		updatedAnnotations = map[string]string{}
```

Comment:
this isn't needed is it?

Reply:
as long as the object has an ObjectMeta, no. `GetAnnotations` could theoretically return nil, since it wraps unstructured.

Reply:
But you're not assigning to `updatedAnnotations`, you're only reading from it, so `nil` and empty are equivalent.

Reply:
but yeah, we shouldn't do weird contortions for unreachable invalid code; updated.